### ISSUE TITLE
PIConGPU on Pascal-based GPUs...

### ISSUE THREAD

**Issue by physicsninja123** [ComputationalRadiationPhysics/picongpu#1562]
Considering the new Pascal-based consumer GPUs like the 1070/1080/Titan GTX have lower FP16 computational power per [http://www.anandtech.com/show/10325/the-nvidia-geforce-gtx-1080-and-1070-founders-edition-review/5] ...

I was wondering if there would be a performance issue using PIConGPU with up to 16 1080 GTXs or Titan Xs, the new Pasdcal-based ones?

Basically, I am looking to use PIConGPU for some ion trap and fusion simulations and I am specing the hardware out now and I would really like to use the new 1080 GTXs, but, I would like to make sure that they are 'computationally compatible' and 'make sense' to use with PIConGPU.

Anyone know if 16bit calculations, as opposed to 32bit ones, or even 64bit, are dominate within PIConGPU, or if there is a way to choose the dominate integer calculation format before compiling?

I thank you very much for your time...

**Comment by ax3l** [comment]
Dear @physicsninja123,

welcome to our repo and thank you for the question!

> I was wondering if there would be a performance issue using PIConGPU with up to 16 1080 GTXs or Titan Xs, the new Pasdcal-based ones?

No of course not and we are looking forward to get our hands on those to do some benchmarks.
Regarding 1080/1070 and "new" Titan X, they are GTX cards meaning they have wonderful single precision flop/s but very few FP64 flops (and as you noted, also less FP16 then Tesla products). Nevertheless, we usually run with single precision (FP32) for most parts of the code and had a great performance with a Titan X (Maxwell, GTX) we got our hands on last year.

FP16 is something we did not experiment with yet (but if you want to, we just need to add a `typedef` and enable it with [this line in the input files](https://github.com/ComputationalRadiationPhysics/picongpu/blob/c7a2fc0bd43c8fffc9e4eee4285595fa79c458b9/src/picongpu/include/simulation_defines/param/precision.param#L29-L35)) with but it might be beneficial in some attributes when using Tesla products, namely P100, as soon as they get available (e.g. on Piz Daint). We should just check our numerical stability with it.

> Basically, I am looking to use PIConGPU for some ion trap and fusion simulations and I am specing the hardware out now and I would really like to use the new 1080 GTXs, but, I would like to make sure that they are 'computationally compatible' and 'make sense' to use with PIConGPU.

The physical applications sound great! Feel free to contact us (group leader: @bussmann) via mail in case you want to more details on the setup and model constrains that you might not want to shout out publicly. (Pls refer to this thread.) I bet you would be interested adding fusion related cross-sections (production rates) and we can certainly assist you to get a feeling about important code details. Also, simulation sizes (2D3V or 3D3V?), target densities, etc. will put constrains on how many cards you will need in terms of RAM.

We personally use only Tesla cards for our simulations due to higher RAM / card (K80) and ECC support (also, ECC is now "free" on Tesla P100). Also the RAM bandwidths are significantly better in the Tesla line, e.g. on Pascal cards it's HBM2 (P100) vs. GDDR5 (1080).
(P.S.: be aware we are soon starting to merge our [multi-platform branch](https://github.com/ComputationalRadiationPhysics/picongpu-alpaka) to our mainline `dev` using [alpaka](https://github.com/ComputationalRadiationPhysics/alpaka) which also supports CPUs where RAM is cheap.)

But to summarize, yes PIConGPU is Pascal-generation "computationally compatible" and "make sense" just depends on the plasma physics details on which we should talk. Nevertheless, in case you already have access to a 1080, 1070 or a DGX-1 lets run a quick test on it first before buying in badges. Just from experience, we never had a hard time transitioning from Tesla to Fermi To Kepler to Maxwell in the past and the Pascal CUDA feature list is wonderful. I expect ourselves to have latest starting with SC16 in November constant access to Pascal cards.

> Anyone know if 16bit calculations, as opposed to 32bit ones, or even 64bit, are dominate within PIConGPU, or if there is a way to choose the dominate integer calculation format before compiling?

Besides a few (live analysis) plugins that are not dominating the runtime, our default setup is nearly 100% FP32 on floating point operations. So explicitly, yes you are good to go with GTX and Tesla product but note the details above.
We normalize all our calculations and all precision options can be set in the above mentioned file called [`precision.param`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/c7a2fc0bd43c8fffc9e4eee4285595fa79c458b9/src/picongpu/include/simulation_defines/param/precision.param#L29-L35).

A last technical note: Pascal cards require CUDA 8+ (currently there is only a 8RC out) and you simply have to compile PIConGPU with C++11 enabled.
